sceptre/resolvers/__init__.py

```diff
@@ -17,37 +17,158 @@
+
+        keys_to_delete = []
+
+        def resolve(containing_list_or_dict, key, obj: Resolver):
+            result = obj.resolve()
+            # If the resolver "resolves to nothing", then it should get deleted out of its container.
+            if result is None:
+                keys_to_delete.append((containing_list_or_dict, key))
+            else:
+                containing_list_or_dict[key] = result
+
+        _call_func_on_values(resolve, self._argument, Resolver)
+        delete_keys_from_containers(keys_to_delete)
+
+        self._argument_is_resolved = True
+
+    def _recursively_setup(self):
+        """Ensures all nested resolvers in this resolver's argument are also setup when this
+        instance's setup method is called.
+        """
+        self.setup()
+
+        def setup_nested(containing_list_or_dict, key, obj: Resolver):
+            obj._recursively_setup()
+
+        _call_func_on_values(setup_nested, self._argument, Resolver)
+
+    def _recursively_clone(self: Self, stack: "stack.Stack") -> Self:
+        """Recursively clones the instance and its arguments.
+
+        The returned instance will have an identical argument that is a different memory reference,
+        so that instances inherited from a stack group and applied across multiple stacks are
+        independent of each other.
+
+        Furthermore, all nested resolvers in this resolver's argument will also be cloned to ensure
+        they themselves are also independent and fully configured for the current stack.
+        """
+
+        def recursively_clone_arguments(obj):
+            if isinstance(obj, Resolver):
+                return obj._recursively_clone(stack)
+            if isinstance(obj, list):
+                return [recursively_clone_arguments(item) for item in obj]
+            elif isinstance(obj, dict):
+                return {
+                    key: recursively_clone_arguments(val) for key, val in obj.items()
+                }
+            return obj
+
+        argument = recursively_clone_arguments(self._argument)
+        clone = type(self)(argument, stack)
+        return clone
+
+    def clone_for_stack(self: Self, stack: "stack.Stack") -> Self:
+        """
+        Obtains a clone of the current object, setup and ready for use for a given Stack instance.
+        """
+        clone = self._recursively_clone(stack)
+        clone._recursively_setup()
+        return clone
+
     def setup(self):
         """
-        This method is called at during stack initialisation.
+        This method is called when the object is connected to a Stack instance.
         Implementation of this method in subclasses can be used to do any
         initial setup of the object.
         """
         pass  # pragma: no cover
 
+    def __str__(self):
+        as_str = f"!{self.__class__.__name__}"
+        if self._argument is not None:
+            as_str += f"({self._argument})"
+
+        return as_str
```

Comment:
This was mostly moved from the placeholders.py file. Previously, placeholders would access `resolver.argument` to represent the resolver when it couldn't be resolved. But if we're going to allow nested resolvers, that won't work. Instead, we need to represent the raw resolvers _all the way down_. Defining a `__str__` method on this base class and using the stringification of `self._argument` ensures that if any resolver in the argument cannot be resolved, it's placeholders end up being used all the way down into its arguments.